Fix wrong drive format on Unix

MojtabaTajik opened the pull request:
Add reading filesystem type from /proc/self/mountinfo on Linux and use sys call as the fallback approach.

- Prioritize reading the filesystem type from `/proc/self/mountinfo` on Linux.
- Retain the existing P/Invoke fallback for non-Linux platforms and as a fallback on Linux.
- Added `try/catch` block to handle potential but unlikely access issues or unrecognized data.
  Although `/proc/self/mountinfo` is process-specific and should generally be accessible,
  this ensures robustness in case of unexpected errors (e.g., temporary unavailability, corrupted data).
- Updated documentation to reflect the reasoning and kernel behaviour regarding procfs on different platforms.

No breaking changes are expected. The refactor should be functionally equivalent but more reliable and efficient.

FIX #95099

MojtabaTajik committed:
Fix wrong drive format on Unix

Change GetFormatInfoForMountPoint method to "use /proc/self/mountinfo" file

- Replaced system calls (`statfs`/`statvfs`) with a simplified parsing of `/proc/self/mountinfo`.
- Removed native SystemNative_GetFormatInfoForMountPoint as we don't need anymore
- Parsing the mount info file allows for efficient retrieval of filesystem format and type.
- No functional changes; the method continues to return the correct filesystem format and drive type for the given mount point.

FIX #95099

MojtabaTajik committed:
Merge branch 'main' into FixWrongDriveFormatOnUnix

am11 commented:
> with a more simplified approach that parses /proc/self/mountinfo to retrieve filesystem information

1. procfs, can be unavailable on linux. It can also be misbehaving and we have made runtime implementation resilient for such cases in number of places using fallbacks.

2. procfs is also optional in FreeBSD. It is not enabled by default.

3. On SunOS et al. procfs is _different_; does not provide binary interface instead of text-based.

4. there is no procfs on macOS.

Given all that, we should remove this common P/Invoke call, but keep it as primary source for non-linux and as a fallback on linux.

am11 commented:
> with a more simplified approach that parses /proc/self/mountinfo to retrieve filesystem information

1. procfs can be unavailable on linux. It can also be misbehaving and we have made runtime implementation resilient for such cases in number of places using fallbacks.

2. procfs is also optional in FreeBSD. It is not enabled by default.

3. On SunOS et al. procfs is _different_; does not provide binary interface instead of text-based.

4. there is no procfs on macOS.

Given all that, we should remove this common P/Invoke call, but keep it as primary source for non-linux and as a fallback on linux.

am11 commented:
> with a more simplified approach that parses /proc/self/mountinfo to retrieve filesystem information

1. procfs can be unavailable on linux. It can also be misbehaving and we have made runtime implementation resilient for such cases in number of places using fallbacks.

2. procfs is also optional in FreeBSD. It is not enabled by default.

3. On SunOS et al. procfs is _different_; provides binary interface instead of text-based.

4. there is no procfs on macOS.

Given all that, we should remove this common P/Invoke call, but keep it as primary source for non-linux and as a fallback on linux.

MojtabaTajik commented:
@am11 Should we go by the "mountinfo" first and go for that statfs syscall if that failed?
In that case, I think checking against "/proc/self/mountinfo" file exists is enough for a fallback mechanism, yes?

am11 commented:
> @am11 Should we go by the "mountinfo" first and go for that statfs syscall if that failed? In that case, I think checking against "/proc/self/mountinfo" file exists is enough for a fallback mechanism, yes?

Problem with File.Exists is that we still may not be able to read the file due to access issues. The reliable way is to actually read the file and TryParse whatever comes back. If it doesn't exist or the data is unavailable or unrecognized, fallback.  We could use one of the existing patterns: https://github.com/dotnet/runtime/issues/62513 (simplest would be: `try/catch { /* ignore */ }`).

> In the case of something like SunOS, where the symlink is represented as a binary format, how should we handle it? We may need to check the operating system or implement a mechanism for the parser method to fail gracefully, triggering a fallback syscall as a recovery option.

We can keep P/Invoking for non-linux:
e.g.
```
#if TARGET_LINUX
try
{
  // read from procf
  // return;
}
catch { /* ignored */ }
#endif

// existing GetFormatInfoForMountPoint logic
```

am11 commented:
> @am11 Should we go by the "mountinfo" first and go for that statfs syscall if that failed? In that case, I think checking against "/proc/self/mountinfo" file exists is enough for a fallback mechanism, yes?

Problem with File.Exists is that we still may not be able to read the file due to access issues. The reliable way is to actually read the file and TryParse whatever comes back. If it doesn't exist or the data is unavailable or unrecognized, fallback.  We could use one of the existing patterns: https://github.com/dotnet/runtime/issues/62513 (simplest would be: `try/catch { /* ignore */ }`).

> In the case of something like SunOS, where the symlink is represented as a binary format, how should we handle it? We may need to check the operating system or implement a mechanism for the parser method to fail gracefully, triggering a fallback syscall as a recovery option.

We can keep P/Invoking for non-linux:
e.g.
```
#if TARGET_LINUX
try
{
  // read from procfs
  // return;
}
catch { /* ignored */ }
#endif

// existing GetFormatInfoForMountPoint logic
```

MojtabaTajik commented:
> Problem with File.Exists is that we still may not be able to read the file due to access issues. 

Based on the [Kernel documentation](https://www.[kernel.org/doc/Documentation/filesystems/proc.txt](https://www.kernel.org/doc/Documentation/filesystems/proc.txt)), There is one such file per process because not all processes see the same mount points.

Since /proc/self/mountinfo refers to the process of reading the file, you should have no issues accessing it within your process. You can simply open and read the file like any other in the /proc directory.

MojtabaTajik committed:
Add support for reading filesystem type from /proc/self/mountinfo on Linux

Add support for reading filesystem type from /proc/self/mountinfo on Linux

- Prioritize reading the filesystem type from `/proc/self/mountinfo` on Linux.
- Retain the existing P/Invoke fallback for non-Linux platforms and as a fallback on Linux.
- Added `try/catch` block to handle potential but unlikely access issues or unrecognized data.
  Although `/proc/self/mountinfo` is process-specific and should generally be accessible,
  this ensures robustness in case of unexpected errors (e.g., temporary unavailability, corrupted data).
- Updated documentation to reflect the reasoning and kernel behavior regarding procfs on different platforms.

am11 commented on src/native/libs/System.Native/entrypoints.c:
```diff
@@ -131,7 +131,7 @@ static const Entry s_sysNative[] =
     DllImportEntry(SystemNative_Realloc)
     DllImportEntry(SystemNative_GetSpaceInfoForMountPoint)
-    DllImportEntry(SystemNative_GetFormatInfoForMountPoint)
+	DllImportEntry(SystemNative_GetFormatInfoForMountPoint)
```
`git checkout HEAD~10 src/native` will undo whitespace changes in this directory.

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -49,9 +49,69 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
             return GetFormatInfoForMountPoint(name, out _, out type);
         }
+		
+		/// <summary>
```
Four space indentation instead of tabs; needed for clean CI.

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -39,7 +39,7 @@ internal static unsafe partial int GetFormatInfoForMountPoint(
             int bufferLength,
             long* formatType);
-
+			
```
nit: undo

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -49,9 +49,69 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+		/// <returns>0 if successful, otherwise an error code.</returns>
+		private static unsafe int GetFormatInfoForMountPoint(string name, out string format, out DriveType type)
+		{
+			#if TARGET_LINUX
```
nit: pre-processor directives are usually placed at the beginning of the line without adhering to current indentation. (same goes to #endif)

am11 commented:
> Btw, I agree that using try-catch is a good approach to handle any unexpected behaviour, as there could be other unknown issues.

Yup, we have actually seen those issues in chroot type environments (qemu, unikernels like nanos (which is the implementation of Linux interface and not the stock kernel), scratch containers).

MojtabaTajik committed:
Fix whitespaces and replace tabs with 4 spaces

MojtabaTajik committed:
Remove pre-processor directives indention

MojtabaTajik committed:
Fix whitespaces

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+        /// <returns>0 if successful, otherwise an error code.</returns>
         private static unsafe int GetFormatInfoForMountPoint(string name, out string format, out DriveType type)
         {
+        #if TARGET_LINUX
```
```suggestion
#if TARGET_LINUX
```

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

MojtabaTajik committed:
fix indention

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+        /// 1. On Linux systems, it first attempts to read from `/proc/self/mountinfo`.
+        /// 2. If step 1 fails or on non-Linux systems, it falls back to a P/Invoke call.
+        ///
+        /// The `/proc/self/mountinfo` approach is preferred on Linux because 
```
```suggestion
        /// The `/proc/self/mountinfo` approach is preferred on Linux because
```

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                }
+            }
+            catch { /* ignored */ }
+#endif
+            
             byte* formatBuffer = stackalloc byte[MountPointFormatBufferSizeInBytes];    // format names should be small
```
```suggestion
#endif

            byte* formatBuffer = stackalloc byte[MountPointFormatBufferSizeInBytes];    // format names should be small
```

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

huoyaoyuan commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+            try
+            {
+                const string mountInfoFilePath = "/proc/self/mountinfo";
+                var mountInfoFileContent = File.ReadAllLines(mountInfoFilePath);
+                foreach (var line in mountInfoFileContent)
+                {
+                    var parser = new StringParser(line, ' ');
```
Style - no `var` unless type name is present

huoyaoyuan commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                    // Get filesystem type (Field 9)
+                    string filesystemType = parser.MoveAndExtractNext();
+            
+                    if (mountPoint.Equals(name, StringComparison.Ordinal))
```
```suggestion
                    if (mountPoint == name)
```
Do not use the instance `Equals` method. `==` uses `Ordinal` behavior.

MojtabaTajik committed:
Use actual type instead of var

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Huo Yaoyuan <huoyaoyuan@hotmail.com>

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,68 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+        /// <returns>0 if successful, otherwise an error code.</returns>
         private static unsafe int GetFormatInfoForMountPoint(string name, out string format, out DriveType type)
         {
+#if TARGET_LINUX
+            try
+            {
+                const string mountInfoFilePath = "/proc/self/mountinfo";
+                var mountInfoFileContent = File.ReadAllLines(mountInfoFilePath);
+                foreach (var line in mountInfoFileContent)
+                {
+                    StringParser parser = new StringParser(line, ' ');
+            
+                    // Skip fields we don't care about (Fields 1-4)
+                    parser.MoveNext(); // Skip Mount ID
+                    parser.MoveNext(); // Skip Parent ID
+                    parser.MoveNext(); // Skip Major:Minor
+                    parser.MoveNext(); // Skip Root
+            
+                    // Get the mount point (Field 5)
+                    string mountPoint = parser.MoveAndExtractNext();
+            
+                    // Skip to the separator which is end of optional fields (Field 8)
+                    while (parser.MoveAndExtractNext() != "-")
+                    {
+                    }
+            
+                    // Get filesystem type (Field 9)
+                    string filesystemType = parser.MoveAndExtractNext();
+            
+                    if (mountPoint == name)
+                    {
+                        format = filesystemType;
+                        type = GetDriveType(filesystemType);
+                        return 0;
+                    }
+                }
+            }
+            catch { /* ignored */ }
+#endif
```
.NET 9 split enumerator can help in making it more memory efficient. Something like this:
```suggestion
#if TARGET_LINUX
            try
            {
                using StreamReader reader = new("/proc/self/mountinfo");
                
                string rawLine;
                while ((rawLine = reader.ReadLine()) is not null)
                {
                    ReadOnlySpan<char> line = rawLine.AsSpan();
                    MemoryExtensions.SpanSplitEnumerator<char> fields = line.Split(' ');
                
                    // Skip fields we don't care about (Fields 1-4)
                    fields.MoveNext(); // Skip Mount ID
                    fields.MoveNext(); // Skip Parent ID
                    fields.MoveNext(); // Skip Major:Minor
                    fields.MoveNext(); // Skip Root

                    if (!MemoryExtensions.Equals(line[fields.Current], name, StringComparison.Ordinal)) continue;

                    // Skip to the separator which is end of optional fields (Field 8)
                    while (fields.MoveNext() && !MemoryExtensions.Equals(line[fields.Current], "-", StringComparison.Ordinal));

                    fields.MoveNext();
                    format = line[fields.Current].ToString();

                    fields.MoveNext();
                    type = GetDriveType(line[fields.Current].ToString());
                    
                    return 0;
            }
            catch { /* ignored */ }
#endif
```

MojtabaTajik replied:
Sorry, I accidentally pressed the accept button in Github. I've reverted the changes.

Your code isn't working correctly because "fields.Current" returns only a single character instead of the entire mount point string.

Btw, I used the string parser based on a suggestion from @tmds, and other similar functionalities use the same class.

am11 replied:
“Something like that” suggestion imply confirming that it’s working before merging. I actually tested it in Linux machine in a standalone app and it was returning the same values before and after.

> Your code isn't working correctly because "fields.Current" returns only a single character instead of the entire mount point string

.Current returns Range and it is only of length 1 for a single character.

MojtabaTajik replied:
I know how it works :)
But in your code, we skip the first 4 fields and start reading, which means we read the root instead of the mount point, I will add your suggestion and push a fix for that. I'm not sure how it works for you! Maybe you can just check against / drive in your test case.

am11 replied:
It was a quick test on wsl with daily build while I was on a tram on my way to work, which is why I was relying on your double-checking. 😅
```c#
Console.WriteLine(GetFormatInfoForMountPoint());

static int GetFormatInfoForMountPoint()
{
    using StreamReader reader = new("/proc/self/mountinfo");

    string rawLine;
    while ((rawLine = reader.ReadLine()) is not null)
    {
        ReadOnlySpan<char> line = rawLine.AsSpan();
        MemoryExtensions.SpanSplitEnumerator<char> fields = line.Split(' ');

        // Skip fields we don't care about (Fields 1-4)
        fields.MoveNext(); // Skip Mount ID
        fields.MoveNext(); // Skip Parent ID
        fields.MoveNext(); // Skip Major:Minor
        fields.MoveNext(); // Skip Root

        string name = "foo";
        //                    if (line[fields.Current] != name) continue;
        //                    fields.MoveNext();
        while (fields.MoveNext() && !MemoryExtensions.Equals(line[fields.Current], "-", StringComparison.Ordinal)) ;// {Console.Wr>                                                                                                                    //MemoryExtens>                                                                                                                    //            >        fields.MoveNext();

        string format = line[fields.Current].ToString();
        Console.WriteLine(format);
        //return 0;
    }

    return 1;
}
```

MojtabaTajik replied:
fine, I pushed the fix for that. also refactored a bit for the check parts.

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

MojtabaTajik committed:
Simplify checks

MojtabaTajik committed:
add new line

MojtabaTajik committed:
fix spaces

MojtabaTajik committed:
Revert "fix spaces"

This reverts commit 5da14ac1c5b955f94e451696d3fed1722fcd0f46.

MojtabaTajik committed:
Revert "add new line"

This reverts commit eba6e78aab3f4f61e36176ec1e99325dab609de7.

MojtabaTajik committed:
Revert "Simplify checks"

This reverts commit c8c0a849ccebde5e615fd627ccbbec46ec505907.

MojtabaTajik committed:
Revert "Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs"

This reverts commit 3505c12666604866995cc409f0cf190273b62503.

MojtabaTajik committed:
Update src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

MojtabaTajik committed:
Skip to mount point

MojtabaTajik committed:
Update comment

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -91,18 +91,19 @@ private static unsafe int GetFormatInfoForMountPoint(string name, out string for
+                    fields.MoveNext(); // Move next to MountPoint field
 
-                    if (!MemoryExtensions.Equals(line[fields.Current], name, StringComparison.Ordinal)) continue;
+                    if (!line[fields.Current].Equals(name, StringComparison.Ordinal)) continue;
```
.Equals is not supported on ROS. We need MemoryExtensions.Equals here.

MojtabaTajik committed:
Use MemoryExtensions.Equals

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                    fields.MoveNext(); // Skip Parent ID
+                    fields.MoveNext(); // Skip Major:Minor
+                    fields.MoveNext(); // Skip Root
+                    fields.MoveNext(); // Skip to MountPoint field
```
This PR should include additional changes so we use the MountPoint field as the Name/RootDirectory for DriveInfo.

tmds replied:
That is: for the instances returned by `DriveInfo.GetDrives`.

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                    format = line[fields.Current].ToString();
+
+                    fields.MoveNext();
+                    type = GetDriveType(line[fields.Current].ToString());
```
Shouldn't we be passing `format` to `GetDriveType` instead of reading the next field?

`man proc` has this example line:
```
36 35 98:0 /mnt1 /mnt2 rw,noatime master:1 - ext3 /dev/root rw,errors=continue
```

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+        /// <param name="type">Output parameter for the drive type.</param>
+        /// <returns>0 if successful, otherwise an error code.</returns>
         private static unsafe int GetFormatInfoForMountPoint(string name, out string format, out DriveType type)
         {
```
It would be nice to have some parsing tests for this, similar to https://github.com/dotnet/runtime/blob/0154a2f3403d94ea6d6f93f5a774b6e366969e0a/src/libraries/Common/tests/Tests/Interop/procfsTests.cs#L36.

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+
+                    return 0;
+            }
+            catch { /* ignored */ }
```
I see this was discussed in https://github.com/dotnet/runtime/pull/107365#issuecomment-2330774825.

Personally, I think the `File.Exists` should be fine, but I defer to the maintainers' preferences.

am11 replied:
The reason why all/most procfs usage in libs are using try-catch pattern is because anything could go wrong (https://github.com/dotnet/runtime/issues/62513). Also see https://github.com/dotnet/runtime/issues/74316#issuecomment-1221685984 and https://github.com/dotnet/runtime/issues/52753 to read on unreliability issues FileSystemInfo.Exists out in the wild.

am11 replied:
The reason why all/most procfs usage in libs are using try-catch pattern is because anything could go wrong (https://github.com/dotnet/runtime/issues/62513). Also see https://github.com/dotnet/runtime/issues/74316#issuecomment-1221685984 and https://github.com/dotnet/runtime/issues/52753 to read on unreliability issues attached with FileSystemInfo.Exists out in the wild.

tmds replied:
There are some reasons why we don't expect these to occur for `/proc/self/mountinfo`

Anyway, my main concern here is that a generic catch all masks failures in the parsing.

Perhaps we can filter for IOException/UnauthorizedAccessException. These are the exceptions we might expect when not being able to read the procfs file.

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                    MemoryExtensions.SpanSplitEnumerator<char> fields = line.Split(' ');
+
+                    // Skip fields we don't care about (Fields 1-4)
+                    fields.MoveNext(); // Skip Mount ID
```
nit: I find the _Skip_ in this comment misleading. The call to `MoveNext` puts us at the Mount ID, not past it, right?

May be you can drop the _Skip_.
```
fields.MoveNext(); // Mount ID
...
```

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+
+                    fields.MoveNext();
+                    type = GetDriveType(line[fields.Current].ToString());
+
```
There doesn't seem to be a matching closing `}` for the `while` loop. If CI doesn't complain, then `TARGET_LINUX` may not be set.

tmds commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+                    type = GetDriveType(line[fields.Current].ToString());
+
+                    return 0;
+            }
```
If we went over the lines of `mountinfo` and didn't find a match, there's no need to do the PInvoke. We can just return immediately.

am11 commented on src/libraries/Common/src/Interop/Unix/System.Native/Interop.MountPoints.FormatInfo.cs:
```diff
@@ -50,8 +50,65 @@ internal static int GetFormatInfoForMountPoint(string name, out DriveType type)
+        /// <returns>0 if successful, otherwise an error code.</returns>
         private static unsafe int GetFormatInfoForMountPoint(string name, out string format, out DriveType type)
         {
+#if TARGET_LINUX
+            try
+            {
+                using StreamReader reader = new("/proc/self/mountinfo");
+
+                string rawLine;
+                while ((rawLine = reader.ReadLine()) is not null)
+                {
+                    ReadOnlySpan<char> line = rawLine.AsSpan();
+                    MemoryExtensions.SpanSplitEnumerator<char> fields = line.Split(' ');
+
+                    // Skip fields we don't care about (Fields 1-4)
+                    fields.MoveNext(); // Skip Mount ID
+                    fields.MoveNext(); // Skip Parent ID
+                    fields.MoveNext(); // Skip Major:Minor
+                    fields.MoveNext(); // Skip Root
+                    fields.MoveNext(); // Skip to MountPoint field
+
+                    if (!MemoryExtensions.Equals(line[fields.Current], name, StringComparison.Ordinal)) continue;
+
+                    // Skip to the separator which is end of optional fields (Field 8)
+                    while (fields.MoveNext() && !MemoryExtensions.Equals(line[fields.Current], "-", StringComparison.Ordinal));
+
+                    fields.MoveNext();
+                    format = line[fields.Current].ToString();
+
+                    fields.MoveNext();
+                    type = GetDriveType(line[fields.Current].ToString());
+
+                    return 0;
+            }
+            catch { /* ignored */ }
+#endif
```
```suggestion
            if (OperatingSystem.IsLinux())
            {
                try
                {
                    using StreamReader reader = new("/proc/self/mountinfo");

                    string? rawLine;
                    while ((rawLine = reader.ReadLine()) is not null)
                    {
                        ReadOnlySpan<char> line = rawLine.AsSpan();
                        MemoryExtensions.SpanSplitEnumerator<char> fields = line.Split(' ');

                        // Skip fields we don't care about (Fields 1-4)
                        fields.MoveNext(); // Skip Mount ID
                        fields.MoveNext(); // Skip Parent ID
                        fields.MoveNext(); // Skip Major:Minor
                        fields.MoveNext(); // Skip Root
                        fields.MoveNext(); // Skip to MountPoint field

                        if (!MemoryExtensions.Equals(line[fields.Current], name, StringComparison.Ordinal)) continue;

                        // Skip to the separator which is end of optional fields (Field 8)
                        while (fields.MoveNext() && !MemoryExtensions.Equals(line[fields.Current], "-", StringComparison.Ordinal));

                        fields.MoveNext();
                        format = line[fields.Current].ToString();

                        fields.MoveNext();
                        type = GetDriveType(line[fields.Current].ToString());

                        return 0;
                    }
                }
                catch { /* ignored */ }
            }
```

I just noticed that this wasn't building on Linux because only CoreLib has TARGET_LINUX type of constants defined. Also, there is a syntax error `try` and `while` blocks have only one closing `}`.

OperatingSystem.IsLinux() block will be optimized out on non-Linux by the JIT (since it returns a constant bool).

tmds replied:
I had noticed the syntax error and based on that assumed TARGET_LINUX must not be set (https://github.com/dotnet/runtime/pull/107365#discussion_r1761081704).

> if (OperatingSystem.IsLinux())

Yes, this is the way!

danmoseley commented:
@MojtabaTajik do you plan to address remaining feedback?

tmds commented:
@MojtabaTajik do you intend to continue working on this?

MojtabaTajik commented:
Sorry for the delay; I’m currently on holiday. I’ll fix it next week.

tmds commented:
@MojtabaTajik is this still on you radar? If not, someone else can pick it up.

tmds commented:
@MojtabaTajik I'd be interested to work on this if it doesn't look like you'll find time for it in the next couple of months. It would be nice to have this in a .NET 10 preview.

tmds commented:
@MojtabaTajik will you be able to work on this so it can be part of a .NET 10 preview? If not, I will have some time for it in the coming weeks.

tmds commented:
@MojtabaTajik I intend to pick this up next week, and I'll close this open PR then.

tmds commented:
Closing in favor of https://github.com/dotnet/runtime/pull/116102.
